TensorArray demo notebook

BryanCutler opened the pull request:
This notebook uses parts of the CoNLL-2003 tutorial to showcase the use of TensorArray to represent BERT embeddings for model training and prediction results.

BryanCutler commented on tutorials/corpus/util.py:
```diff
@@ -73,6 +74,8 @@ def predict_on_df(df: pd.DataFrame, id_to_class: Dict[int, str], predictor):
     result_df["predicted_iob"] = iobs
     result_df["predicted_type"] = types
+    prob_values = predictor.predict_proba(x_values)
+    result_df["predicted_class_pr"] = tp.TensorArray(prob_values)
```
I couldn't use this function from my notebook, I just made a copy of it. Is it something we want to keep updated with this?

BryanCutler replied:
TODO: The notebook version is slightly different now

frreiss replied:
Having multiple copies of this function is ok. If this code was core functionality, I would have put this function into the library proper instead of leaving it in a directory of notebooks.

BryanCutler commented:
cc @frreiss to please take a look, also @kmh4321 and @MLnick if interested.

BryanCutler commented:
@frreiss I had a problem in the notebook trying to align a span column with the token_id of the BERT DataFrame. The span column gets dropped with the DataFrame is saved, but I wanted to show something at the end, but I seem to be missing something to add back in the span column.

MLnick commented:
@BryanCutler where is the notebook itself? Doesn't seem part of the PR currently.

BryanCutler commented:
Oops, sorry @MLnick , it would help if I added the file. Doing it now..

frreiss reviewed (changes requested):
Still not seeing the notebook itself in these commits. Did you forget to add a file?

frreiss commented:
Ok, now the commit with the notebook is showing up. Strange. Will review momentarily.

frreiss commented on notebooks/Model_Training_with_TensorArray.ipynb:
```diff
@@ -0,0 +1,5826 @@
+    }
+   ],
+   "source": [
+    "train_results_df = predict_on_df(train_df, int_to_label, base_model)\n",
```
I don't think it makes much sense to run inference on the training set. I recommend you skip ahead to running inference on the test set.

frreiss commented on notebooks/Model_Training_with_TensorArray.ipynb:
```diff
@@ -0,0 +1,5826 @@
+   ],
+   "source": [
+    "# Look at results on the test set\n",
+    "test_results_df = predict_on_df(corpus_df[corpus_df[\"fold\"] == \"test\"], \n",
```
These results contain label probabilities for each of the BERT tokens in the test fold of the corpus. To obtain predictions of the named entities in the test fold, you need to convert these probabilities to one "best" label (IOB tag + entity type) per token. The easiest way to get those "best" labels is to take the highest probability label at each token position. Then you 'll want to convert those token labels to (span, entity type pairs), the same way that you processed the corpus itself earlier in this notebook. Once you've made that conversion, you can compute precision and recall a la Analyze_Model_Outputs.ipynb.

frreiss reviewed (approved):
This is looking good. Could use some additional explanations at the beginning and some additional postprocessing/scoring code at the end (see comments throughout), but this is in good enough shape to merge as-is for now.

BryanCutler commented:
Thanks for reviewing @frreiss , I'll work on an update

BryanCutler commented on notebooks/Model_Training_with_TensorArray.ipynb:
```diff
@@ -0,0 +1,5826 @@
+    "\n",
+    "## TensorArray Example Usage\n",
+    "\n",
+    "To demostrate the TensorArray functionality from Text Extensions for Pandas, we will use the CoNLL-2003 dataset [TODO REF] to \n",
```
Thanks @frreiss , I'll add that. Should I reference the other more complete CoNLL tutorial notebooks we have also?

frreiss replied:
The stuff in the `tutorials` directory is currently a bit half-baked (also somewhat subject to anonymity constraints). I recommend including a reference to the other CoNLL-related notebook in this directory, though -- `Analyze_Model_Outputs.ipynb`

MLnick commented on notebooks/Model_Training_with_TensorArray.ipynb:
```diff
@@ -0,0 +1,5826 @@
+{
```
<p>"Standard arithetic and comparison ..." -&gt; "Standard <strong>arithmetic</strong> and comparison ..."</p><p> </p><p>"... elements of similar shaprs." -&gt; "...  elements of similar shapes."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

MLnick replied:
<p>In <code>spans_df = spans_df.reindex(columns=mns=["span", "token_span", "ent_type"])</code>  is it supposed to be <code>columns=mns=...</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

MLnick commented:
Just a couple minor things picked up so far - still working through the technical details of the notebook. Though agreed with @frreiss that some more description of the problem & goal etc would be useful.

BryanCutler commented:
Thanks @MLnick , I will reword things to better state the problem & goal.

BryanCutler commented:
@frreiss and @MLnick , I updated a lot of the wording and added more descriptions and links. Hopefully, this better states the problem and goal of the notebook. Please take another look when you can, thanks!

BryanCutler commented on text_extensions_for_pandas/array/test_tensor.py:
```diff
@@ -345,12 +345,35 @@ def test_sum(self):
+    def test_take(self):
+        # Test no missing gets same dtype
+
+        # Test missing with nan fill gets promoted to float
```
TODO

BryanCutler commented on text_extensions_for_pandas/io/bert.py:
```diff
@@ -216,7 +216,7 @@ def align_bert_tokens_to_corpus_tokens(
             .rename(columns={"corpus_token": "span"})
     )
-    cons_df["span"] = TokenSpanArray.align_to_tokens(
+    cons_df["token_span"] = TokenSpanArray.align_to_tokens(
```
@frreiss before this change it is replacing the existing "span" column, could you just verify that is what we want or should we do this to make a new "token_span" column?

BryanCutler replied:
this is also the cause of the current test failure

frreiss replied:
Returning both a token-based span and the original makes sense. Would you mind updating the description of the returned DataFrame's schema in the docstring above to match? If I'm reading this code correctly, the schema should be `[span, token_span, ent_type]`, where `span` is the original span from the input and `token_span` is a token-based span aligned to the BERT tokens.

BryanCutler replied:
On second thought, adding an additional column "token_span" makes the notebook a bit more confusing and later functions are set to work with a "span" column, so it would need to be changed too.

I revert my changes so that the "span" column is replaced by a TokenSpanArray, as before, and updated the docs. I think it's clearer this way, wdyt?

BryanCutler commented:
Not sure why the ReviewNB link wasn't added, but here it is

https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/

frreiss reviewed (commented):
Code changes outside the notebook look good.

frreiss commented on notebooks/Model_Training_with_BERT.ipynb:
```diff
@@ -0,0 +1,5371 @@
+{
```
<p>This cell can be merged with the next cell. Also, you can move the initialization of <code>bert</code>  down to the later cell where that object is first used.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

frreiss replied:
<p>I recommend moving the code that initializes the variable <code>bert</code>  to this cell, so the reader doesn't have to scroll up to find it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

frreiss replied:
<p>Can you move the text of this comment to the markdown cell above and correct the spelling of "convenience"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

frreiss commented on notebooks/Model_Training_with_BERT.ipynb:
```diff
@@ -0,0 +1,5371 @@
+{
```
<p>Can you add a computation of precision and recall at this point in the notebook, then redo the precision/recall computation after correcting for the BERT model splitting tokens? There should be a measurable increase in accuracy as a result of that correction.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

BryanCutler replied:
<p>Yeah, that's a good idea. It does give it a little bump ~0.85 -&gt; ~0.87 for F1. @frreiss do you happen to know why the model sometimes splits tokens that way - is it an expected artifact or is something wrong?</p>

frreiss commented on notebooks/Model_Training_with_BERT.ipynb:
```diff
@@ -0,0 +1,5371 @@
+{
```
<p>Take a look at Analyze_Model_Outputs.ipynb for an example of drilling down to problematic documents based on document-level precision and recall numbers.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

BryanCutler replied:
I'll try to cover this as a followup

BryanCutler commented on text_extensions_for_pandas/array/tensor.py:
```diff
@@ -224,14 +224,24 @@ def take(
+
+            # Check if there are missing indices to fill, if not can use numpy take below
+            has_missing = np.any(indices < 0)
+            if has_missing:
```
fyi, need this check so that it doesn't unnecessarily promote ints to floats if there are no missing values. Noticed this when sorting a DataFrame with TensorArray, which is now in a cell of the notebook in https://github.com/CODAIT/text-extensions-for-pandas/pull/122
